util/mmio.c

```diff
@@ -6,6 +6,77 @@
 #include <pthread.h>
 #include <stdbool.h>
 
+#ifdef __s390x__
+#include <sys/auxv.h>
+#include <ccan/minmax.h>
+
+bool s390_is_mio_supported;
+
+static __attribute__((constructor)) void check_mio_supported(void)
+{
+#ifdef HWCAP_S390_PCI_MIO
+	s390_is_mio_supported = !!(getauxval(AT_HWCAP) & HWCAP_S390_PCI_MIO);
+#elif
+	s390_is_mio_supported = 0;
+#endif
+}
+
+typedef void (*mmio_memcpy_x64_fn_t)(void *, const void *, size_t);
+/* This uses the STT_GNU_IFUNC extension to have the dynamic linker select the
+   best above implementations at runtime. */
+#if HAVE_FUNC_ATTRIBUTE_IFUNC
+void mmio_memcpy_x64(void *, const void *, size_t)
+	__attribute__((ifunc("resolve_mmio_memcpy_x64")));
+static mmio_memcpy_x64_fn_t resolve_mmio_memcpy_x64(uint64_t);
+#else
+__asm__(".type mmio_memcpy_x64, %gnu_indirect_function");
+write64_fn_t resolve_mmio_memcpy_64(uint64_t)
+	__asm__("mmio_memcpy_x64");
+#endif
+
+#define S390_MAX_WRITE_SIZE 128
+#define S390_BOUNDARY_SIZE (1 << 12)
+#define S390_BOUNDARY_MASK (S390_BOUNDARY_SIZE - 1)
+
+static uint8_t get_max_write_size(void *dst, size_t len)
+{
+	size_t offset = ((uint64_t __force)dst) & S390_BOUNDARY_MASK;
+	size_t size = min_t(int, len, S390_MAX_WRITE_SIZE);
+
+	if (likely(offset + size <= S390_BOUNDARY_SIZE))
+		return size;
+
+	return S390_BOUNDARY_SIZE - offset;
+}
+
+static void mmio_memcpy_x64_mio(void *dst, const void *src, size_t bytecnt)
+{
+	size_t size;
+
+	/* Input is 8 byte aligned 64 byte chunks. The alignment matches the
+	 * requirements of pcistbi but we must not cross a 4K byte boundary.
+	 */
+	while (bytecnt > 0) {
+		size = get_max_write_size(dst, bytecnt);
+		if (size > 8)
+			s390_pcistbi(dst, src, size);
+		else
+			s390_pcistgi(dst, *(uint64_t *)src, 8);
+		src += size;
+		dst += size;
+		bytecnt -= size;
+	}
+}
+
+mmio_memcpy_x64_fn_t resolve_mmio_memcpy_x64(uint64_t hwcap)
+{
+	if (hwcap & HWCAP_S390_PCI_MIO)
```

Comment:
Sorry, just noticed I completely forgot the check if `HWCAP_S390_PCI_MIO` is defined here. Will push a fixed version shortly

Reply:
Should be resolved in the latest version.